enterprise/server/raft/driver/driver.go

```diff
@@ -0,0 +1,435 @@
+	gossiped := false
+	for start := 0; start < numReplicas; start += batchSize {
+		nu := &rfpb.NodeUsage{Nhid: rsp.GetNode().GetNhid()}
+		end := start + batchSize
+		if end > numReplicas {
+			end = numReplicas
+		}
+		for _, rr := range rsp.GetRangeReplicas()[start:end] {
+			nu.ReplicaUsage = append(nu.ReplicaUsage, rr.GetReplicaUsage())
+		}
+		buf, err := proto.Marshal(nu)
+		if err != nil {
+			return err
+		}
+		if err := d.gossipManager.SendUserEvent(constants.NodeUsageEvent, buf, false /*=coalesce*/); err != nil {
+			return err
+		}
+		gossiped = true
+	}
+
+	// If a node does not yet have any replicas, the loop above will not
+	// have gossiped anything. In that case, gossip an "empty" usage event
+	// now.
+	if !gossiped {
+		nu := &rfpb.NodeUsage{Nhid: rsp.GetNode().GetNhid()}
+		buf, err := proto.Marshal(nu)
+		if err != nil {
+			return err
+		}
+		if err := d.gossipManager.SendUserEvent(constants.NodeUsageEvent, buf, false /*=coalesce*/); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (d *Driver) OnEvent(updateType serf.EventType, event serf.Event) {
+	switch updateType {
+	case serf.EventUser:
+		userEvent, _ := event.(serf.UserEvent)
+		d.handleEvent(&userEvent)
+	default:
+		break
+	}
+}
+
+func (d *Driver) handleEvent(event *serf.UserEvent) {
+	if event.Name != constants.NodeUsageEvent {
+		return
+	}
+	nu := &rfpb.NodeUsage{}
+	if err := proto.Unmarshal(event.Payload, nu); err != nil {
+		return
+	}
+	if nu.GetNhid() == "" {
+		log.Warningf("Ignoring malformed driver node usage: %+v", nu)
+		return
+	}
+	d.clusterMap.ObserveNode(nu.GetNhid())
+	for _, ru := range nu.GetReplicaUsage() {
+		d.clusterMap.ObserveReplica(nu.GetNhid(), ru)
+	}
+	atomic.AddInt64(&d.numSamples, 1)
+}
+
+func (d *Driver) manageClusters() error {
+	if ns := atomic.LoadInt64(&d.numSamples); ns < 5 {
+		log.Debugf("Not managing cluster yet, sample count: %d", ns)
+		return nil
+	}
+	ctx := context.Background()
+	rsp, err := d.store.ListCluster(ctx, &rfpb.ListClusterRequest{
+		LeasedOnly: true,
+	})
+	if err != nil {
+		return err
+	}
+
+	leasedClusterIDs := make([]uint64, 0, len(rsp.GetRangeReplicas()))
+	for _, rr := range rsp.GetRangeReplicas() {
+		for _, replica := range rr.GetRange().GetReplicas() {
+			leasedClusterIDs = append(leasedClusterIDs, replica.GetClusterId())
+			break
+		}
+	}
+	deadReplicas := d.clusterMap.DeadReplicas(leasedClusterIDs, d.opts.ReplicaTimeout)
+	log.Printf("Dead replicas: %+v, my clusters: %+v", deadReplicas, leasedClusterIDs)
+	// AddClusterNode()
+
+	overloadedReplicas := d.clusterMap.OverloadedReplicas(leasedClusterIDs, d.opts.MaxReplicaSizeBytes)
+	log.Printf("Overloaded replicas: %+v, my clusters: %+v", overloadedReplicas, leasedClusterIDs)
+	// SplitCluster()
+
+	excessReplicas := d.clusterMap.ExcessReplicas(leasedClusterIDs, rsp.GetNode().GetNhid())
+	log.Printf("Excess replicas: %+v, my clusters: %+v", excessReplicas, leasedClusterIDs)
+	// AddClusterNode && RemoveClusterNode
```

Comment:
intentional? and above

Reply:
yeah in a followup that's where I'm going to actually call those methods, just is a reminder for me.